drivers/serial/uart_cdns.c

```diff
@@ -0,0 +1,299 @@
+{
+	struct uart_cdns_regs *uart_regs = DEV_UART(dev);
+
+	int i = 0;
+
+	for (i = 0; i < size && (!uart_cdns_is_rx_fifo_empty(uart_regs)); i++) {
+		rx_data[i] = uart_regs->rx_tx_fifo;
+	}
+	if (i > 0) {
+		uart_regs->ctrl |= CTRL_RSTTO_MASK;
+	}
+	return i;
+}
+
+void uart_cdns_enable_tx_irq(const struct device *dev)
+{
+	struct uart_cdns_regs *uart_regs = DEV_UART(dev);
+
+	uart_regs->intr_enable |= CSR_TTRIG_MASK;
+}
+
+void uart_cdns_disable_tx_irq(const struct device *dev)
+{
+	struct uart_cdns_regs *uart_regs = DEV_UART(dev);
+
+	uart_regs->intr_disable |= CSR_TTRIG_MASK;
+}
+
+static int uart_cdns_irq_tx_ready(const struct device *dev)
+{
+	return !uart_cdns_is_tx_fifo_full(DEV_UART(dev));
+}
+
+void uart_cdns_enable_rx_irq(const struct device *dev)
+{
+	struct uart_cdns_regs *uart_regs = DEV_UART(dev);
+
+	uart_regs->rx_timeout = DEFAULT_RTO_PERIODS_FACTOR;
+	uart_regs->intr_enable |= (CSR_RTRIG_MASK | CSR_RBRK_MASK | CSR_TOUT_MASK);
+}
+
+/** Disable RX UART interrupt */
+void uart_cdns_disable_rx_irq(const struct device *dev)
+{
+	struct uart_cdns_regs *uart_regs = DEV_UART(dev);
+
+	uart_regs->intr_disable |= (CSR_RTRIG_MASK | CSR_RBRK_MASK | CSR_TOUT_MASK);
+}
+
+static int uart_cdns_irq_rx_ready(const struct device *dev)
+{
+	return !uart_cdns_is_rx_fifo_empty(DEV_UART(dev));
+}
+
+static void uart_cdns_enable_irq_err(const struct device *dev)
+{
+	struct uart_cdns_regs *uart_regs = DEV_UART(dev);
+
+	uart_regs->intr_enable |=
+		(CSR_TOVR_MASK | CSR_TOUT_MASK | CSR_PARE_MASK | CSR_FRAME_MASK | CSR_ROVR_MASK);
+}
+
+static void uart_cdns_disable_irq_err(const struct device *dev)
+{
+	struct uart_cdns_regs *uart_regs = DEV_UART(dev);
+
+	uart_regs->intr_disable |=
+		(CSR_TOVR_MASK | CSR_TOUT_MASK | CSR_PARE_MASK | CSR_FRAME_MASK | CSR_ROVR_MASK);
+}
+
+static int uart_cdns_is_irq_pending(const struct device *dev)
+{
+	struct uart_cdns_regs *uart_regs = DEV_UART(dev);
+
+	return (uart_regs->channel_intr_status != 0);
+}
+
+/** Check for IRQ updates */
+static int uart_cdns_update_irq(const struct device *dev)
+{
+	return 1;
+}
+
+/** Set the callback function pointer for IRQ. */
+void uart_cdns_set_irq_callback(const struct device *dev, uart_irq_callback_user_data_t cb,
+				   void *cb_data)
+{
+	struct uart_cdns_data *data;
+
+	data = DEV_DATA(dev);
+	data->callback = cb;
+	data->cb_data = cb_data;
+}
+
+static void uart_cdns_irq_handler(const struct device *dev)
+{
+	struct uart_cdns_regs *uart_regs = DEV_UART(dev);
+	uint32_t key = irq_lock();
+	uint32_t isr_status;
+	struct uart_cdns_data *data = DEV_DATA(dev);
+
+	if (data->callback) {
+		data->callback(dev, data->cb_data);
+	}
+
+	/* clear events */
+	/* need to make local copy of the status */
+	isr_status = uart_regs->channel_intr_status;
+
+	irq_unlock(key);
+}
+
+#endif
+
+static const struct uart_driver_api uart_cdns_driver_api = {
+	.poll_in = uart_cdns_poll_in,
+	.poll_out = uart_cdns_poll_out,
+	.err_check = NULL,
```

Comment:
I don't see `.configure`, `.config_get` with `CONFIG_UART_USE_RUNTIME_CONFIGURE`, I think these would be possible for this IP..... but it doesn't have to come with this PR

Reply:
Yes, will add them later.